### ISSUE TITLE
playing new sound will play a already paused/stopped sound prev

### ISSUE THREAD

**Issue by littleblack111** [wwmm/easyeffects#3422]
### EasyEffects Version

7.1.8

### What package are you using?

Arch (easyeffects)

### Distribution

Arch BTW

### Describe the bug

currently: https://github.com/wwmm/easyeffects/issues/3386

but theres also two anomaly:
- if i stop spotify, now i play sth like youtube. it would play like the last 0.1s of the spotify again then continue playing youtube
- rn my temp solution to #3386 is do a while true loop for autostart.
    - the issue is when it's starting, it would glitch my speaker for a spit second, its some what fine, until... today. i have 2 players on, i just paused spotify(and it crashed) # THEN IT JUMPSCARD ME SOO HARD. it just give like a LARGE 5000% sound, like a beep. then it returned to normal...

### Expected Behavior

none of above happening

### Debug Log

<details><summary>Debug Log</summary><br/>

```
Paste your log here
```
</details>

### Additional Information

https://github.com/wwmm/easyeffects/issues/3386

**Comment by wwmm** [comment]
I think we already have a discussion about this somewhere. But I can't remember the issue name. In any case it is a hard to solve this as things are right now on PipeWire's side. There isn't a pipeline management infrastructure like the one provided by GStreamer. Where you can force the pipeline to remain active until the last plugin has received a message sent by the first plugin. PipeWire is in total control of the plugins activation  as well as for how long they keep running.

>the issue is when it's starting, it would glitch my speaker for a spit second, its some what fine, until... today. i have 2 players on, i just paused spotify(and it crashed) THEN IT JUMPSCARD ME SOO HARD. it just give like a LARGE 5000% sound, like a beep. then it returned to normal...

I wonder if you soundcard drier has the same problem as mine. When my soundcard comes back from the suspended state it does a very annoying crackling. It has nothing to do with EasyEffects. The difference is that depending on the plugins I am using the crackling amplitude can be very high. But it is still there when EasyEffects is not running.

In cases like this the only solution is disabling the soundcard suspension in the server configuration.

**Comment by wwmm** [comment]
Based on the list of plugins you provided on the other issue EasyEffects pipeline should have almost no data stored inside each plugin. In other words the extra latency displayed in EasyEffects bottom panel should be very small and virtually unnoticeable by ear.

**Comment by littleblack111** [comment]
no. but if my volume is around 80-90 and im playing smth loud, it would crack rly hard too...

and idk wats up but my speaker performance VERY different between linux and windows(i dual boot)

in windows, 10% of it is around 80% of it in linux. and in windows it never crack(volume)...

its a eden

**Comment by wwmm** [comment]
>In other words the extra latency displayed in EasyEffects bottom panel should be very small and virtually unnoticeable by ear.

Actually this isn't the case when using the `loudness` plugin. I tested it now and it adds quite a lot of latency depending on the FFT size. With the default `4096` value the extra pipeline latency jumps to more than 90 milliseconds. Does the glitch changes if you select smaller FFT sizes?

**Comment by wwmm** [comment]
Forget my last comment. You are using the `Bass Loudness` and not the `loudness` plugin...

**Comment by wwmm** [comment]
With the plugins you are using the extra latency should be close to zero. What means that the amount of data stored in it is negligible. I wonder from where the "last 0.1s sound" you are listening to is coming from.

**Comment by littleblack111** [comment]
no idea... but def sth dealing with easyeffects because its nowhere to be seen if i kill easyeffects

**Comment by wwmm** [comment]
> no idea... but def sth dealing with easyeffects because its nowhere to be seen if i kill easyeffects

I wonder if the one storing old buffers is the null-sink... If issue is still there when EE global bypass is enabled or the plugins are removed from the pipeline then the null-sink is the one with the old data.

**Comment by ManuVice** [comment]
any news here? I have the same problem with an USB DAC

**Comment by wwmm** [comment]
> any news here? I have the same problem with an USB DAC

The situation is still the same on my side. Somehow on my hardware I can't notice this happening. My guess is that the interaction between pipewire and the soundcard drivers are relevant. When I am listening to something on spotify and then play a youtube video I do not listen a fraction of the paused spotify song like described in the first post.

**Comment by ManuVice** [comment]
> I do not listen a fraction of the paused spotify song like described in the first post.

Thats exactly my problem. A fraction from the last played media is played.
I will take a look at it. Thank you

**Comment by wwmm** [comment]
@ManuVice which plugins are you using? What is the quantum and rate values pw-top shows for your soundcard when you playing something?

**Comment by ManuVice** [comment]
Iam running 1. multiband-compressor 2. crystalizer 3. equalizer 4. limiter

**Spotify playing:**
![spotify](https://github.com/user-attachments/assets/e7f86260-7a19-4ad9-9433-c9f7cd5b1583)

**Spotify closed**
![spotify closed](https://github.com/user-attachments/assets/aaee0eaf-12d9-4197-8c21-4a948191c93c)

**Open Youtube frontpage without playing any video it plays a short sound like a "plopp"**
![open youtube](https://github.com/user-attachments/assets/035d5cb9-db11-44a0-9635-3b075f07b1ff)

**Playing a Youtube video**
![playing youtube](https://github.com/user-attachments/assets/88edb2df-ecd4-4c58-83b6-ea8adca11bc0)

**Going back from Youtube video to Youtube frontpage**
![stop youtube video](https://github.com/user-attachments/assets/af4454c8-381f-44c8-a77d-835c32970ca1)

**Comment by wwmm** [comment]
A quantum of 2048 combined with a sampling rate of 48000 Hz would lead to a minimum latency of about `2048 / 48000 = 0.043` seconds. This is more or less what is going to happen in the vast majority of users installations. Including mine. IT does not seem big enough to be causing this issue.

>Open Youtube frontpage without playing any video it plays a short sound like a "plopp"

Is this "plopp" really an audio that is part of what was being played before or just an weird noise? Maybe what is happening is the noise some soundcards do when PipeWire takes them from the suspended state back to the active state. My soundcard does that. EasyEffects effects just make the noise more noticeable.

Another possibility is that you have set the equalizer to a mode that is not `FIR`. The other ones add a lot of latency. What also means previous audio data stored inside the plugin. This has happened to some users in the past.

**Comment by ManuVice** [comment]
equalizer is set to IIR

> Is this "plopp" really an audio that is part of what was being played before or just an weird noise?

90% of the time I hear a small cut from my last played track or a voice from the last played video


I didnt touch my pipewire config. Only things I have changed are: 

- pipewire-pulse.conf -> "resample.quality = 10"
- /etc/wireplumber/wireplumber.conf.d/50-alsa-config.conf
- /etc/wireplumber/wireplumber.conf.d/51-disable-suspension.conf


51-disable-suspension.conf:

- copied from arch linux wiki

in 50-alsa-config.conf:

-  api.alsa.period-size   = 256
-  api.alsa.headroom      = 0

I played a lot with these two settings (api.alsa.period-size, api.alsa.headroom) and quantum settings in pipewire.conf but nothing changed for me. I rolled back the most settings to default

**Comment by DaviDagostin** [comment]
This problem happens to me too. Setting the equalizer to IIR generates an imperceptible peak on resume and practically solves the problem.
Another possibility would be a fade in filter of a few milliseconds every time an audio stream starts.

**Comment by wwmm** [comment]
>Another possibility would be a fade in filter of a few milliseconds every time an audio stream starts.

I wonder if it is possible to do this. We do not have control over the start and stop operation. PipeWire is the one controlling that. It may seem strange but it is the way things are handled. We only link the filters and handle inside each one of them the audio buffers that come. The filter does not know if somewhere far away a stream is starting.

We can separately monitor each stream state and get this information. But as the whole process is asynchronous PipeWire may have already resumed the flow of buffers to each plugin by the time we get to know if it is a new stream the one responsible tor the new buffers.

What I think is odd is that some users had this problem in the past even when we still provided a control that forced the output effects pipeline to remain active for several seconds. In a situation like this one all the old buffers should have already been drained. That is why I suspect that is most cases what is happening is the noise some cards generate when PipeWire brings them back from suspended state.

>Setting the equalizer to IIR generates an imperceptible peak on resume and practically solves the problem.

For example let's consider the case of an user that uses only the equalizer. The pipeline would be 
```
player -> null-sink -> equalizer -> spectrum -> output_level -> soundcard
```
. With the equalizer in the IIR mode the additional latency reported by the plugin is zero. So there should be no data stored inside it. The spectrum and the output level meter work in passthrough mode. So they can not be the ones responsible for old data. So the one left to point the finger to would be the null-sink used to create the virtual device. But this is 100% PipeWire's code. We just ask it to load the virtual device for us.

Maybe the null-sink is storing data. What would be a little unexpected.

**Comment by wwmm** [comment]
>If I switch the tab in gnome settings I have audio distortion. I wanted to record the whole thing with obs-studio, but if I start obs the problem is gone.. I had to record it with my phone.. If I disable easyeffects or start obs no distortion.
With this mobile phone video you can unfortunately only hear very poorly my bad..

For some reason the videos can not be played. But based on what your are describing the issue is probably related to PipeWire's dynamic latency switching. What you are doing on GNOME settings is probably making PipeWire switch to a different quantum value. Try to take a look at `pw-top` output. Depending on which plugins you have enabled on EasyEffects this can cause noises because some plugins still need to be reset when the latency changes.

The fact that keeping OBS opened fixes it is also understandable for the same reasons. The latency requested by OBS is probably lower than the one requested by gnome settings. So the switching does not happen while OBS is running.

**Comment by ManuVice** [comment]
May I ask which sound card you are using?

**Comment by wwmm** [comment]
> May I ask which sound card you are using?

The one from the motherboard https://rog.asus.com/motherboards/rog-strix/rog-strix-b650e-e-gaming-wifi-model/helpdesk_bios/. I use its optical port output.

**Comment by Safari77** [comment]
```
mpv pinknoise.wav
sleep 3
mpv silence.wav
```
When starting `silence.wav`, there is 148 ms of audio from pinknoise at the beginning. With equalizer on the delay is 148 ms.
If I keep playing `silence.wav` constantly at the background, I don't get extra noises AT ALL, but easyeffects is using CPU all the time (3%, but electricity is cheap :) ).

**Comment by wwmm** [comment]
>With equalizer on the delay is 148 ms.

That is a lot of delay. Are you using the equalizer in `FIR` mode or any of the modes different from `IIR`? They will store data inside the plugin and increase latency. As well as causing the issue described here. This is expected. What is stragne is that some people are having this while using the IIR mode.

**Comment by wwmm** [comment]
> there is 148 ms of audio from pinknoise at the beginning.

Oh... It is not a delay. It is just the length of the audio data. Are you able to put these files here?

**Comment by Safari77** [comment]
```
sox -n -c 2 -r 48000 -b 16 pink_noise.wav synth 10 pinknoise gain -12
sox -n -c 2 -r 48000 -b 16 silence.wav trim 0 10
```

**Comment by wwmm** [comment]
> ```
> sox -n -c 2 -r 48000 -b 16 pink_noise.wav synth 10 pinknoise gain -12
> sox -n -c 2 -r 48000 -b 16 silence.wav trim 0 10
> ```

I am doing some tests and the results depend on the plugin. So far things are fine with the `bass enhancer`, `exciter` and the `equalizer` in `IIR` mode. But I can hear a fraction of the old data when the `limiter`, the `convolver` or the `crystalizer` are in the pipeline.

>If I do PEQ in pipewire, I don't get delays or garbage sounds.

The ideal solution would be to make the pipeline keep playing for some time after the last player stops sending buffers to our virtual device. it is probably what PipeWire is doing in its built-in filter chain. The problem is that it does not offer a way for us to do the same. Or at least it didn't the last time I looked at the available API.

**Comment by wwmm** [comment]
I have the feeling that it will be more efficient to revisit this issue once the port to Qt is ready. There I am doing again something we used to do in the past and it definitely helps the QT branch to deal with it. But it won't help the current gtk branch as it is. 

The inactivity timeout in EasyEffects preferences window used to be applied to both pipelines instead of just the microphone pipeline. But as relinking our filters on the fly seems to trigger a PipeWire bug that makes one of the channels to not have audio I've stopped unlinking and relinking the output effects pipeline on demand. But in the Qt branch I've decided to bring back this procedure although it is not clear if PipeWire has fixed the other bug on their side.

The reason why this procedure helps the Qt branch to deal with this old data issue is that there we are able to actually destroy the filter class after the disconnection. What clears the internal data. For some reason I could never figure out this caused crashes in the gtk branch.

**Comment by Safari77** [comment]
I verified that if I use `http://lsp-plug.in/plugins/ladspa/para_equalizer_x32_lr`  in **FIR** Equalizer mode as `libpipewire-module-filter-chain` ladspa plugin (when EasyEffects is **NOT** running) I also get the ~150 ms sound from old buffers.

**Comment by theofficialgman** [comment]
Same issue here. amd (first gen) framework laptop 13 using the "Gracefu's Edits" profile from here https://github.com/cab404/framework-dsp

**Comment by KekYamaha** [comment]
I have the same issue here on a laptop with an AMD CPU, I don't know if it's related to the bug but it's worth mentioning. The bug doesn't occur when EE is not running.
Running this command with EE active:

>pactl list sinks | grep -A3 "State"

State: SUSPENDED
Name: alsa_output.pci-0000_66_00.6.analog-stereo
Description: Ryzen HD Audio Controller Analog Stereo
Driver: PipeWire


State: SUSPENDED
Name: easyeffects_sink
Description: Easy Effects Sink
Driver: PipeWire

while audio is playing shows State: RUNNING, after pausing it's IDLE and after 3 seconds it goes to SUSPENDED and that's when I resume playing audio I hear a small portion of previous sound for a split second. I even tried disabling all active plugins and the issue persists.

**Comment by wwmm** [comment]
> while audio is playing shows State: RUNNING, after pausing it's IDLE and after 3 seconds it goes to SUSPENDED and that's when I resume playing audio I hear a small portion of previous sound for a split second. I even tried disabling all active plugins and the issue persists.

@KekYamaha which plugins are you using? Which EasyEffects version?

**Comment by KekYamaha** [comment]
@wwmm Hello! I am using easyeffects 8.0.9 and I only use Equalizer and Stereo Tools, however on a fresh profile without any plugins active the bug is still there.

**Comment by wwmm** [comment]
>Hello! I am using easyeffects 8.0.9 and I only use Equalizer and Stereo Tools, however on a fresh profile without any plugins active the bug is still there.

Old audio present even without plugins is very strange. The only possible explanation that comes to my mind at this moment is that the old data is stored inside pipewire's null-sink. But I honestly do not remember if the null-sink is supposed to do that.

**Comment by wiedaar** [comment]
Same thing here, it started a week or 2 ago out of nothing.
When playing a new song a little bit of the song before is heard.

I am running cachyos.

I had this problem before on mint or fedora, i forgot
i am on version 8.1.0

When i turn off al the effects accept the EQ the problem is gone.
When i turn on the other effects one by one and playing in between them for testing, it went from a pop sound to a little bit of the sound played from the audio before that.

I have tried switching the effects in line, but that didn't work.

i have the limiter, (mine is in dutch) hardheid and hardheid van bas, and vertraging (delay) but i only use that when im playing on my bluetooth speakers also

**Comment by wwmm** [comment]
> i have the limiter, (mine is in dutch) hardheid and hardheid van bas, and vertraging (delay) but i only use that when im playing on my bluetooth speakers also

@wiedaar did you set a readahead value in the limiter? This mode probably buffers data inside the plugin.

**Comment by wiedaar** [comment]
> > i have the limiter, (mine is in dutch) hardheid and hardheid van bas, and vertraging (delay) but i only use that when im playing on my bluetooth speakers also
> 
> [@wiedaar](https://github.com/wiedaar) did you set a readahead value in the limiter? This mode probably buffers data inside the plugin.

i left the limiter standard only turned automatic level on

what i found out,  only EQ everthing is ok, turn loudness on and there is the bit of music before the new music. And when i turn on the limiter also, youtube movies are out of sync the first seconds

i took me days to find out what te problem was, trying a different audio output, via my nvidia card in stead of my sound card, turning hardware acceleration off, tried a new browser, i also was thinking of installing a different distro, tried all the suggestions

but now the sollution :)

**Comment by wwmm** [comment]
> turn loudness on and there is the bit of music before the new music

Depending on how the loudness plugin is configured it can introduce a lot of extra latency. Same thing with the equalizer. And a lot of extra latency often means old data stored inside the plugin.

**Comment by Digitalone1** [comment]
> Depending on how the loudness plugin is configured it can introduce a lot of extra latency. Same thing with the equalizer. And a lot of extra latency often means old data stored inside the plugin.

Sometimes it happens to me also.

Can we do something to reduce this side effect? Maybe playing silence for the amount of latency after all the streams are closed?

**Comment by wwmm** [comment]
> Sometimes it happens to me also.

I think that the equalizer only saves a lot of data internally if in a mode that is not `IIR`. Are you doing that? In the IIR mode there should be nothing noticeable there.

>Can we do something to reduce this side effect? Maybe playing silence for the amount of latency after all the streams are closed?

It is not impossible. One approach could be extending the test signal class so it has a silent mode. And somehow set its state to active when no stream is playing to EasyEffects. We already force it to play to our virtual sink https://github.com/wwmm/easyeffects/blob/a0715615818af58a2534818d5175ce81de9e922b/src/test_signals.cpp#L266.

I will try to remember to take a look at this.

**Comment by wwmm** [comment]
Instead of trying to get the current extra latency value it will probably be easier to use the `activity timeout`. Besides being a configurable value it is larger. So in theory it should guarantee a full drain.

**Comment by Safari77** [comment]
try this:
make this file `~/.config/wireplumber/main.lua.d/50-alsa-config.lua` with contents of:
```
rule = {
  matches = {
    {
      { "node.name", "matches", "alsa_output.*" },
    },
  },
  apply_properties = {
    ["session.suspend-timeout-seconds"] = 0,
  },
}
```
then `systemctl --user daemon-reload ; systemctl --user restart pipewire{,.socket} pipewire-pulse{,.socket} wireplumber ; pkill easyeffects`

**Comment by Digitalone1** [comment]
> I think that the equalizer only saves a lot of data internally if in a mode that is not `IIR`. Are you doing that? In the IIR mode there should be nothing noticeable there.

Yes, I use the default mode, but I think this issue is introduced by some other plugins...

**Comment by wiedaar** [comment]
> try this: make this file `~/.config/wireplumber/main.lua.d/50-alsa-config.lua` with contents of:
> 
> ```
> rule = {
>   matches = {
>     {
>       { "node.name", "matches", "alsa_output.*" },
>     },
>   },
>   apply_properties = {
>     ["session.suspend-timeout-seconds"] = 0,
>   },
> }
> ```
> 
> then `systemctl --user daemon-reload ; systemctl --user restart pipewire{,.socket} pipewire-pulse{,.socket} wireplumber ; pkill easyeffects`

does that fix the latency problem?

**Comment by wwmm** [comment]
>["session.suspend-timeout-seconds"] = 0,

Although this will help to confirm if playing silence will help avoiding the hardware suspension permanently will waste cpu power.

**Comment by wiedaar** [comment]
for the time being i'm only using the EQ,good to see that you are working on it wwmm :)

**Comment by wwmm** [comment]
> for the time being i'm only using the EQ,good to see that you are working on it wwmm :)

At this moment I am trying to improve the new `voice suppressor` plugin. But I will try to do the test signal hack after it.

**Comment by Safari77** [comment]
> > then `systemctl --user daemon-reload ; systemctl --user restart pipewire{,.socket} pipewire-pulse{,.socket} wireplumber ; pkill easyeffects`
> 
> does that fix the latency problem?

Latency is not what this issue #3422 is about.

**Comment by Safari77** [comment]
> > ["session.suspend-timeout-seconds"] = 0,
> 
> Although this will help to confirm if playing silence will help avoiding the hardware suspension permanently will waste cpu power.

What waste? Top says 2% gnome-shell, then 1% firefox... are there configurations where this causes increased CPU usage?

**Comment by wwmm** [comment]
> What waste? Top says 2% gnome-shell, then 1% firefox... are there configurations where this causes increased CPU usage?

It is not waste as in "my CPU is at 100%". It is more about the audio server discretely playing silence in situations where it could have stopped. What can propagate to easyeffects too depending on the activity timeout value. Eventually easyeffects will try to unlink its pipeline. What avoids it wasting cpu power. But when the soundcard is allowed to suspend pipewire will stop the filters sooner.

For people on desktops it does not matter much. But for laptops running on battery it is something to keep in mind.

**Comment by wwmm** [comment]
>But when the soundcard is allowed to suspend pipewire will stop the filters sooner.

Or at least that is the expectation. We still force our pipeline to be unlinked because I have seen pipewire failing to stop the filters in some corner cases. Like when we use easyeffects echo canceller.

**Comment by wwmm** [comment]
I am looking at this right now and there is a difficulty I did not consider before. When the test signal plays to our virtual sink it triggers the same link signals we use to decide when to link or unlink the filters. As a result easyeffects enters a loop where the test signal that flushes the pipeline is enabled and disabled infinitely. I am not sure about the best way to deal with this yet...

But something else I remembered now is that when easyeffects unlinks the filters it also destroys them. So even if the filters are saving seconds of data internally it is irrelevant. The data is destroyed as soon as the inactivity timeout triggers.

As I have been suspecting for a while PipeWire's null-sink must be storing data inside them. I wonder if there is a way to make it not do that.

**Comment by wwmm** [comment]
>But something else I remembered now is that when easyeffects unlinks the filters it also destroys them. So even if the filters are saving seconds of data internally it is irrelevant. The data is destroyed as soon as the inactivity timeout triggers.

It is what I thought was happening. But after further investigation it doesn't seem to be the case. Hum...

**Comment by wwmm** [comment]
>It is what I thought was happening. But after further investigation it doesn't seem to be the case. Hum...

What actually happens is that we destroy the filter if it is not in the plugin list anymore. For example if the convolver is in the pipeline and a preset that does not have the it but has all the other active plugins is loaded the convolver is destroyed. What definitely will throw away old data.

 The question is if it is ok to destroy all plugins by default. There is definitely a performance penalty in unnecessarily recreating the plugins class all the time...

**Comment by wwmm** [comment]
>As I have been suspecting for a while PipeWire's null-sink must be storing data inside them. I wonder if there is a way to make it not do that.

After several tests it does not seem to be the case. The old data I was listening was from the plugins that were not destroyed for the reasons above. So either we somehow avoid the infinity test signal loop or we force all plugins to be recreated whenever the pipeline resumes operation.

**Comment by Digitalone1** [comment]
> So either we somehow avoid the infinity test signal loop or we force all plugins to be recreated whenever the pipeline resumes operation.

I think the second would bring other issues and performance problems, so I believe the first would be better.

**Comment by wwmm** [comment]
> I think the second would bring other issues and performance problems, so I believe the first would be better.

Probably. Another issue is that if the backend is destroyed and and the qml window isn't it is possible qml will try to use invalid pointers.

**Comment by wwmm** [comment]
@Digitalone1 the approaches above do not seem viable. So I did something else. The plugins now will have a `clear_data()` method they have to override from the base class. This method is called when the pipeline is unlinked. In the latest master branch update I implemented it for a few plugins like the equalizer, the convolver and the limiter. Although not all plugins need this it is probably better to implement for the others too.

As clear_data is only called when the pipeline is unlinked the activity timeout is directly related to when the clearing happens.

**Comment by Digitalone1** [comment]
Okay, but you implemented 

```c++
{
  std::scoped_lock<std::mutex> lock(data_mutex);

  lv2_wrapper->destroy_instance();
}
```

Only for some plugins. Shouldn't we do it for all?

And a similar process for the LADSPA wrapper? 

Update: Okay, now I understood. You didn't add that part to Convolver and Crystalizer because they do not use the LV2 wrapper. But I suppose we need something similar for the LADSPA wrapper.

**Comment by wwmm** [comment]
> Update: Okay, now I understood. You didn't add that part to Convolver and Crystalizer because they do not use the LV2 wrapper. But I suppose we need something similar for the LADSPA wrapper.

Yes. For the ones that use LV2 it is just copy and paste. The others will need something different. But it should still be enough to call setup after it gets small modifications.

**Comment by Digitalone1** [comment]
> Yes. For the ones that use LV2 it is just copy and paste. The others will need something different. But it should still be enough to call setup after it gets small modifications.

I'm doing it.

Should I add also

```c++
if (rate == 0 || n_samples == 0) {  
  // Some signals may be emitted before PipeWire calls our setup function
  return;
}
```

On top of `setup()` for every plugin?

**Comment by wwmm** [comment]
> > Yes. For the ones that use LV2 it is just copy and paste. The others will need something different. But it should still be enough to call setup after it gets small modifications.
> 
> I'm doing it.
> 
> Should I add also
> 
> if (rate == 0 || n_samples == 0) {  
>   // Some signals may be emitted before PipeWire calls our setup function
>   return;
> }
> 
> On top of `setup()` for every plugin?

Yes. Now that pipewire's thread isn't the only calling `setup()` it is safer to insert this test in all of them. Otherwise we have the risk of calling it with rate or n_samples values that are equal to zero.

**Comment by Digitalone1** [comment]
Okay, for the Deepfilternet it was after the scoped_lock, but I'm moving it on top.

The clear_data for this plugin needs also a destroy_instance implementation for LADSPA wrapper. I left a TODO comment.

**Comment by wwmm** [comment]
> The clear_data for this plugin needs also a destroy_instance implementation for LADSPA wrapper. I left a TODO comment.

I am looking at this plugin current setup code and it may not be necessary to modify the wrapper. THe whole approach is different in this plugin. In its setup function we have

```
if (ladspa_wrapper->has_instance()) {
          ladspa_wrapper->deactivate();

          ladspa_wrapper = std::make_unique<ladspa::LadspaWrapper>("libdeep_filter_ladspa.so", "deep_filter_stereo");
        }
```
. So whenever setup is called the `libdeep_filter_ladspa` instance is recreated. We could just call the setup function. But it will be better to do something like

```
void DeepFilterNet::clear_data() {

if (ladspa_wrapper->has_instance()) {
          ladspa_wrapper->deactivate();

          ladspa_wrapper = std::make_unique<ladspa::LadspaWrapper>("libdeep_filter_ladspa.so", "deep_filter_stereo");
        }

setup();

}
```
. The reason is that now I see this setup function does not have the optimization the ones related to LV2 have that avoids reinitialization when pipewire changes just n_samples instead of the rate. So in this case it is better do explicitly put the clear data instructions inside clear_data().

**Comment by Digitalone1** [comment]
You mean this, right?

```c++
void DeepFilterNet::clear_data() {
  if (ladspa_wrapper->has_instance()) {
      ladspa_wrapper->deactivate();

      ladspa_wrapper = std::make_unique<ladspa::LadspaWrapper>("libdeep_filter_ladspa.so", "deep_filter_stereo");
  }
}
```

Without the last call to `setup()`.

**Comment by wwmm** [comment]
> You mean this, right?
> 
> void DeepFilterNet::clear_data() {
>   if (ladspa_wrapper->has_instance()) {
>       ladspa_wrapper->deactivate();
> 
>       ladspa_wrapper = std::make_unique<ladspa::LadspaWrapper>("libdeep_filter_ladspa.so", "deep_filter_stereo");
>   }
> }
> 
> Without the last call to `setup()`.

I had thought with the setup function there. But thinking about it again it probably is not needed.

**Comment by wwmm** [comment]
>I had thought with the setup function there. But thinking about it again it probably is not needed.

No. It is safer to call the setup function. If we recreate the wrapper without the calls
```
ladspa_wrapper->n_samples = n_samples;
ladspa_wrapper->create_instance(48000);
```
something bad may happen.

**Comment by Digitalone1** [comment]
So, since

```
  if (ladspa_wrapper->has_instance()) {
      ladspa_wrapper->deactivate();

      ladspa_wrapper = std::make_unique<ladspa::LadspaWrapper>("libdeep_filter_ladspa.so", "deep_filter_stereo");
  }
```

is already done in `setup`, I suppose we just need to call setup in `clear_data`, right?

**Comment by wwmm** [comment]
> is already done in `setup`, I suppose we just need to call setup in `clear_data`, right?

It will work. But it will probably make it harder avoiding calling setup when n_samples changes. Just like the LV2 plugins this one probably does not need a reset whenever pipewire changes latency.

If we just call setup and add the logic to avoid calls when n_sample changes we may get a situation where `clear_data` won´t do anything.

**Comment by Digitalone1** [comment]
Okay, but unfortunately this is not working. I'm doing some tests and it's crashing every time I remove a plugin from the pipeline.

The coredump says it's crashing here:

```
#0  0x00005557d03fe500 in lv2::Lv2Wrapper::create_instance (this=0x0, rate=@0x5557ec8c00d4: 1)
    at .../src/lv2_wrapper.cpp:239
239	  if (instance != nullptr && this->rate == rate) {
``` 

I'll make the PR so you can test it on your own with all the changes I made.

**Update**: the crash only happens when I'm reproducing something. If I remove a plugin when there's no reproduction, nothing happens.

**Update 2**: I'm playing a song with the DeepFilterNet and now I'm seeing some lines like the following on the terminal.

```
2026-01-21T22:23:01.027Z | WARN |  df::tract | Possible clipping detected (1.029).
```

What the hell does that mean? 😅

**Comment by wwmm** [comment]
>Update: the crash only happens when I'm reproducing something. If I remove a plugin when there's no reproduction, nothing happens.

I saw this before your PR. I am still investigate the reason why this is happening.

>Update 2: I'm playing a song with the DeepFilterNet and now I'm seeing some lines like the following on the terminal.

The deepfilternet library prints things like this from time to time 😄

**Comment by wwmm** [comment]
First round of fixes in the master branch. It seems fine now. The source of the problem was subtle. But it was not the first time we faced a situation like this. The problem was caused by 

```
QMetaObject::invokeMethod(
      QApplication::instance(),
      [this] {
        lv2_wrapper->create_instance(rate);

        std::scoped_lock<std::mutex> lock(data_mutex);

        ready = true;
      },
      Qt::QueuedConnection);
```
. Replacing `QApplication::instance()` by `baseWorker` fixed the crash. At first sight it seemd ok to have `QApplication::instance()` there and running the callback in the main thread. But in this particular case `baseWorker` has an advantage. Its lifetime is bound to the class lifetime. So the callback it runs will run while the class is alive. The main thread was trying to do it when the class was already dead.

**Comment by ManuVice** [comment]
Thanks for looking into this. I have updated to v8.1.1 and for now I couldnt reproduce any old sound artifacts. Seems you have fixed it :-)

**Comment by wwmm** [comment]
> Thanks for looking into this. I have updated to v8.1.1 and for now I couldnt reproduce any old sound artifacts. Seems you have fixed it :-)

Thanks for the feedback. I guess that now we have to do the same for all plugins.

**Comment by Digitalone1** [comment]
> Thanks for the feedback. I guess that now we have to do the same for all plugins.

Didn't we do it already? 🤔

**Comment by wwmm** [comment]
> Didn't we do it already? 🤔

For some reason I thought there were still some plugins needing this. But as far as I can see it seems completed indeed.

**Comment by littleblack111** [comment]
okay, then i will close the issue unless there are still plugins with the issue
